# Issue 8785: avoid subtle interaction between importing multiprocessing and twisted

archive/issues_008785.json:
```json
{
    "assignees": [
        "https://github.com/jasongrout"
    ],
    "body": "<div id=\"comment:0\"></div>\n\nIt turns out that on some platforms, importing multiprocessing, then twisted, leads to an \"int object is not callable\" TypeError.  This breaks devel/sage/sage/all.py's quit_sage function, causing a big traceback at exit.   This could also cause great confusion for people writing a program that uses `@`parallel('multiprocessing') followed by anything involving twisted. \n\nA simple fix is to import the relevant part of twisted before using multiprocessing in `@`parallel.   The attached patch does this.\n\nNOTE: The system that exhibits this is in a corporate setting, and no devs have systems where this can be replicated at present, unfortunately.  So please do NOT revert this just because you don't see the problem on your laptop!\n\nComponent: **misc**\n\nAuthor: **William Stein**\n\nReviewer: **Mariah Lennox**\n\nMerged: **sage-4.4.1.alpha0**\n\n_Issue created by migration from https://trac.sagemath.org/ticket/8785_\n\n",
    "closed_at": "2010-04-28T17:36:19Z",
    "created_at": "2010-04-27T20:54:37Z",
    "labels": [
        "https://github.com/sagemath/sage/labels/p%3A%20major%20/%203",
        "https://github.com/sagemath/sage/labels/bug"
    ],
    "milestone": "https://github.com/sagemath/sage/milestones/sage-4.4.1",
    "reactions": [],
    "repository": "https://github.com/sagemath/sage",
    "title": "avoid subtle interaction between importing multiprocessing and twisted",
    "type": "issue",
    "updated_at": "2016-01-10T23:21:51Z",
    "url": "https://github.com/sagemath/sage/issues/8785",
    "user": "https://github.com/williamstein"
}
```
<div id="comment:0"></div>

It turns out that on some platforms, importing multiprocessing, then twisted, leads to an "int object is not callable" TypeError.  This breaks devel/sage/sage/all.py's quit_sage function, causing a big traceback at exit.   This could also cause great confusion for people writing a program that uses `@`parallel('multiprocessing') followed by anything involving twisted. 

A simple fix is to import the relevant part of twisted before using multiprocessing in `@`parallel.   The attached patch does this.

NOTE: The system that exhibits this is in a corporate setting, and no devs have systems where this can be replicated at present, unfortunately.  So please do NOT revert this just because you don't see the problem on your laptop!

Component: **misc**

Author: **William Stein**

Reviewer: **Mariah Lennox**

Merged: **sage-4.4.1.alpha0**

_Issue created by migration from https://trac.sagemath.org/ticket/8785_





---

archive/issue_events_110116.json:
```json
{
    "actor": "https://github.com/williamstein",
    "created_at": "2010-04-27T20:54:37Z",
    "event": "milestoned",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "milestone_number": null,
    "milestone_title": "sage-4.4.1",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110116"
}
```



---

archive/issue_events_110117.json:
```json
{
    "actor": "https://github.com/williamstein",
    "created_at": "2010-04-27T20:54:37Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "label": "https://github.com/sagemath/sage/labels/p%3A%20major%20/%203",
    "label_color": "ffbb00",
    "label_name": "p: major / 3",
    "label_text_color": "ffffff",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110117"
}
```



---

archive/issue_events_110118.json:
```json
{
    "actor": "https://github.com/williamstein",
    "created_at": "2010-04-27T20:54:37Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "label": "https://github.com/sagemath/sage/labels/bug",
    "label_color": "d73a4a",
    "label_name": "bug",
    "label_text_color": "ffffff",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110118"
}
```



---

archive/issue_events_110119.json:
```json
{
    "actor": "https://github.com/jasongrout",
    "created_at": "2010-04-27T20:54:37Z",
    "event": "assigned",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "subject": "https://github.com/williamstein",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110119"
}
```



---

archive/issue_comments_071429.json:
```json
{
    "body": "Description changed:\n``````diff\n--- \n+++ \n@@ -1,3 +1,5 @@\n It turns out that on some platforms, importing multiprocessing, then twisted, leads to an \"int object is not callable\" TypeError.  This breaks devel/sage/sage/all.py's quit_sage function, causing a big traceback at exit.   This could also cause great confusion for people writing a program that uses `@`parallel('multiprocessing') followed by anything involving twisted. \n \n A simple fix is to import the relevant part of twisted before using multiprocessing in `@`parallel.   The attached patch does this.\n+\n+NOTE: The system that exhibits this is in a corporate setting, and no devs have systems where this can be replicated at present, unfortunately.  So please do NOT revert this just because you don't see the problem on your laptop!\n``````\n",
    "created_at": "2010-04-27T20:58:20Z",
    "formatter": "markdown",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/8785#issuecomment-71429",
    "user": "https://github.com/williamstein"
}
```

Description changed:
``````diff
--- 
+++ 
@@ -1,3 +1,5 @@
 It turns out that on some platforms, importing multiprocessing, then twisted, leads to an "int object is not callable" TypeError.  This breaks devel/sage/sage/all.py's quit_sage function, causing a big traceback at exit.   This could also cause great confusion for people writing a program that uses `@`parallel('multiprocessing') followed by anything involving twisted. 
 
 A simple fix is to import the relevant part of twisted before using multiprocessing in `@`parallel.   The attached patch does this.
+
+NOTE: The system that exhibits this is in a corporate setting, and no devs have systems where this can be replicated at present, unfortunately.  So please do NOT revert this just because you don't see the problem on your laptop!
``````




---

archive/issue_comments_071430.json:
```json
{
    "body": "<div id=\"comment:2\" align=\"right\">comment:2</div>\n\nAttachment: **[trac_8785.patch.gz](https://github.com/sagemath/sage/files/ticket8785/trac_8785.patch.gz)**",
    "created_at": "2010-04-27T20:58:45Z",
    "formatter": "markdown",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/8785#issuecomment-71430",
    "user": "https://github.com/williamstein"
}
```

<div id="comment:2" align="right">comment:2</div>

Attachment: **[trac_8785.patch.gz](https://github.com/sagemath/sage/files/ticket8785/trac_8785.patch.gz)**



---

archive/issue_events_110120.json:
```json
{
    "actor": "https://github.com/williamstein",
    "created_at": "2010-04-27T20:58:45Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "label": "https://github.com/sagemath/sage/labels/needs%20review",
    "label_color": "7fff00",
    "label_name": "needs review",
    "label_text_color": "ffffff",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110120"
}
```



---

archive/issue_events_110121.json:
```json
{
    "actor": "https://github.com/sagetrac-mariah",
    "created_at": "2010-04-27T22:23:45Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "label": "https://github.com/sagemath/sage/labels/needs%20review",
    "label_color": "7fff00",
    "label_name": "needs review",
    "label_text_color": "ffffff",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110121"
}
```



---

archive/issue_events_110122.json:
```json
{
    "actor": "https://github.com/sagetrac-mariah",
    "created_at": "2010-04-27T22:23:45Z",
    "event": "labeled",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "label": "https://github.com/sagemath/sage/labels/positive%20review",
    "label_color": "dfffc0",
    "label_name": "positive review",
    "label_text_color": "ffffff",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110122"
}
```



---

archive/issue_comments_071431.json:
```json
{
    "body": "<div id=\"comment:4\" align=\"right\">comment:4</div>\n\nI tried it and it works!",
    "created_at": "2010-04-27T22:27:20Z",
    "formatter": "markdown",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/8785#issuecomment-71431",
    "user": "https://github.com/sagetrac-mariah"
}
```

<div id="comment:4" align="right">comment:4</div>

I tried it and it works!



---

archive/issue_events_110123.json:
```json
{
    "actor": "https://github.com/williamstein",
    "created_at": "2010-04-28T17:36:19Z",
    "event": "unlabeled",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "label": "https://github.com/sagemath/sage/labels/positive%20review",
    "label_color": "dfffc0",
    "label_name": "positive review",
    "label_text_color": "ffffff",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110123"
}
```



---

archive/issue_events_110124.json:
```json
{
    "actor": "https://github.com/williamstein",
    "created_at": "2010-04-28T17:36:19Z",
    "event": "closed",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_event",
    "url": "https://github.com/sagemath/sage/issues/8785#event-110124"
}
```



---

archive/issue_comments_071432.json:
```json
{
    "body": "Merged: **4.4.1.alpha0**",
    "created_at": "2010-04-28T17:36:19Z",
    "formatter": "markdown",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/8785#issuecomment-71432",
    "user": "https://github.com/williamstein"
}
```

Merged: **4.4.1.alpha0**



---

archive/issue_comments_071433.json:
```json
{
    "body": "Reviewer: **Mariah Lennox**",
    "created_at": "2010-04-29T17:04:28Z",
    "formatter": "markdown",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/8785#issuecomment-71433",
    "user": "https://github.com/sagetrac-mvngu"
}
```

Reviewer: **Mariah Lennox**



---

archive/issue_comments_071434.json:
```json
{
    "body": "Author: **William Stein**",
    "created_at": "2010-04-29T17:04:28Z",
    "formatter": "markdown",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/8785#issuecomment-71434",
    "user": "https://github.com/sagetrac-mvngu"
}
```

Author: **William Stein**



---

archive/issue_comments_071435.json:
```json
{
    "body": "Changed merged from **4.4.1.alpha0** to **sage-4.4.1.alpha0**",
    "created_at": "2010-05-03T01:08:40Z",
    "formatter": "markdown",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/8785#issuecomment-71435",
    "user": "https://github.com/sagetrac-mvngu"
}
```

Changed merged from **4.4.1.alpha0** to **sage-4.4.1.alpha0**



---

archive/issue_comments_071436.json:
```json
{
    "body": "<div id=\"comment:8\" align=\"right\">comment:8</div>\n\nWould be more informative to write explicitely on which hardware/OS it failed. \"corporate settings\" is more than vague. Was there any upstream report? This problem might have been solved since then!",
    "created_at": "2016-01-10T23:21:51Z",
    "formatter": "markdown",
    "issue": "https://github.com/sagemath/sage/issues/8785",
    "type": "issue_comment",
    "url": "https://github.com/sagemath/sage/issues/8785#issuecomment-71436",
    "user": "https://github.com/videlec"
}
```

<div id="comment:8" align="right">comment:8</div>

Would be more informative to write explicitely on which hardware/OS it failed. "corporate settings" is more than vague. Was there any upstream report? This problem might have been solved since then!
